Detectors/TPC/calibration/SpacePoints/src/TrackInterpolation.cxx

```diff
@@ -0,0 +1,427 @@
+    return false;
+  }
+  //float ca, sa;
+  //o2::utils::sincosf(clTOFAlpha, sa, ca);
+  //float clTOFX = clTOF.getX() * ca + clTOF.getY() * sa;                                 // cluster x in sector coordinate frame
+  //std::array<float, 2> clTOFYZ{ -clTOF.getX() * sa + clTOF.getY() * ca, clTOF.getZ() }; // cluster y and z in sector coordinate frame
+  float clTOFX = clTOF.getX();
+  std::array<float, 2> clTOFYZ{ clTOF.getY(), clTOF.getZ() };
+  std::array<float, 3> clTOFCov{ mSigYZ2TOF, 0.f, mSigYZ2TOF }; // assume no correlation between y and z and equal cluster error sigma^2 = (3cm)^2 / 12
+  if (!propagator->PropagateToXBxByBz(trkWork, clTOFX, o2::constants::physics::MassPionCharged, mMaxSnp, mMaxStep, mMatCorr)) {
+    LOG(debug) << "Failed final propagation to TOF radius";
+    return false;
+  }
+  if (!trkWork.update(clTOFYZ, clTOFCov)) {
+    LOG(debug) << "Failed to update extrapolated ITS track with TOF cluster";
+    return false;
+  }
+
+  // go back through the TPC and store updated track positions
+  for (int iRow = param::NPadRows; iRow--;) {
+    if (!mCache->clAvailable[iRow]) {
+      continue;
+    }
+    if (!trkWork.rotate(mCache->clAngle[iRow])) {
+      LOG(debug) << "Failed to rotate track during back propagation";
+      return false;
+    }
+    if (!propagator->PropagateToXBxByBz(trkWork, param::RowX[iRow], o2::constants::physics::MassPionCharged, mMaxSnp, mMaxStep, mMatCorr)) {
+      LOG(debug) << "Failed on back propagation";
+      //printf("trkX(%.2f), clX(%.2f), clY(%.2f), clZ(%.2f), alphaTOF(%.2f)\n", trkWork.getX(), param::RowX[iRow], clTOFYZ[0], clTOFYZ[1], clTOFAlpha);
+      return false;
+    }
+    mCache->posInt[iRow][0] = trkWork.getY();
+    mCache->posInt[iRow][1] = trkWork.getZ();
+    mCache->covInt[iRow][0] = trkWork.getSigmaY2();
+    mCache->covInt[iRow][1] = trkWork.getSigmaZY();
+    mCache->covInt[iRow][2] = trkWork.getSigmaZ2();
+  }
+
+  // calculate weighted mean at each pad row (assume for now y and z are uncorrelated) and store residuals to TPC clusters
+  unsigned short deltaRow = 0;
+  unsigned short nMeasurements = 0;
+  for (int iRow = 0; iRow < param::NPadRows; ++iRow) {
+    if (!mCache->clAvailable[iRow]) {
+      ++deltaRow;
+      continue;
+    }
+    float wTotY = 1.f / mCache->covExt[iRow][0] + 1.f / mCache->covInt[iRow][0];
+    float wTotZ = 1.f / mCache->covExt[iRow][2] + 1.f / mCache->covInt[iRow][2];
+    mCache->pos[iRow][0] = (mCache->posExt[iRow][0] / mCache->covExt[iRow][0] + mCache->posInt[iRow][0] / mCache->covInt[iRow][0]) / wTotY;
+    mCache->pos[iRow][1] = (mCache->posExt[iRow][1] / mCache->covExt[iRow][2] + mCache->posInt[iRow][1] / mCache->covInt[iRow][2]) / wTotZ;
+
+    TPCClusterResiduals res;
+    res.setDY(mCache->clYZ[iRow][0] - mCache->pos[iRow][0]);
+    res.setDZ(mCache->clYZ[iRow][1] - mCache->pos[iRow][1]);
+    res.setY(mCache->pos[iRow][0]);
+    res.setZ(mCache->pos[iRow][1]);
+    res.setPhi(mCache->phi[iRow]);
```

Comment:
I think it is better to use the average (no need for weighting) also for the phi and tgl, rather than using ITS interpolation only

Reply:
Done